lib/server.js

```diff
@@ -305,23 +318,54 @@ CoAPServer.prototype._handle = function(packet, rsinfo) {
   request.rsinfo = rsinfo
   response.statusCode = '2.05'
   response._request = request._packet
+  response._cachekey = toCacheKey(rsinfo.address, rsinfo.port, packet)
 
   var block2cache = this._block2Cache;
+  var block2cache_expiry = this._options.block2CacheExpiryMs;
   //inject this function so the response can add an entry to the cache
   response._addCacheEntry = function(key, payload) {
-    debug('add payload to cache, key:', key);
-    block2cache.key = key;
-    block2cache.payload = payload;
+    if (block2cache.hasOwnProperty(key)) {
+      debug('delete old cache entry, key:', key)
+      clearTimeout(block2cache[key].timeoutId)
+      delete block2cache[key]
+    }
+
+    debug('add payload to cache, key:', key)
+    block2cache[key] = {'payload': payload}
+
+    // setup clean-up timer
+    block2cache[key].timeoutId = setTimeout(function(k) {
+      debug('delete expired cache entry, key:', k)
+      delete block2cache[k]
+    }, block2cache_expiry, key)
```

Comment:
on this comment, instead of

```js
  // setup new expiry timer
    block2cache[key].timeoutId = setTimeout(function(k) {
      debug('delete expired cache entry, key:', k)
      delete block2cache[k]
    }, block2cache_expiry, key)
```

```js
  // setup new expiry timer
    block2cache[key].timeoutId = setTimeout(expire, block2cache_expiry, key)

// somewhere else in the file, at the top level
function expire (block2cache, k) {
      debug('delete expired cache entry, key:', k)
      delete block2cache[k]
} 
```